cherry pick  #100341 #98576 #95836 on 1.18 to enable SSA with APIService

kevindelgado opened the pull request:
Cherry pick of #100341 #98576 #95836 on release-1.18.

#100341: Add ability to skip OpenAPI handler installation
* The [line](https://github.com/kubernetes/kubernetes/commit/66d2f4359efadd2cf85f9d901676f2ea28555951#diff-e7ff48774cfb26a502cd0486130dd008784e6ce27fe81f16a9b534786ae103e8R145-R147) that skips open API Installation in the kube-aggregator has been removed because OpenAPI generation does not exist in kube-aggregator until 1.19.

#98576: Declare TCP default for service port protocol
* The [commit](https://github.com/kubernetes/kubernetes/pull/98576/commits/dab511224020f6f7a896560d7c48770187a683c1) declaring TCP as the default protocol has been removed because it uses features implemented in #96480 that do not exist in 1.18.
* A [line](https://github.com/kubernetes/kubernetes/pull/100715/commits/5ebc98dfb9011ea6d80fcfa692e8e6bf37d06428#diff-1083ad1d288464507552098a30e5f816d56c19f15edf36aaae46afdf43fe1ecdR47) has been added to set the protocol of the service object in the etcd data stub in order to account for SSA's inability to set a default protocol in 1.18

#95836: Add yaml util to unmarshal numbers into int/float 

fixes #99046

```release-note
Fixes using server-side apply with APIService resources
```

nodo committed:
Add yaml util to unmarshal numbers into int/float

liggitt committed:
Stop clearing OpenAPIConfig for kube-aggregator

liggitt committed:
Stop skipping APIService in apply test

liggitt committed:
Use apply to create objects in TestApplyStatus

kevindelgado committed:
Add ability to skip OpenAPI handler installation

kevindelgado commented:
Note that the ability to skip OpenAPI handler installation in the standalone kube-aggregator was removed from this cherry-pick compared to the 1.19 and 1.20 cherry-pick(just this one [line](https://github.com/kubernetes/kubernetes/pull/100713/files#diff-e7ff48774cfb26a502cd0486130dd008784e6ce27fe81f16a9b534786ae103e8R145-R147)) because OpenAPI generation was erroneously missing from kube-aggregator until 1.19 (and fixed via these two PRs (https://github.com/kubernetes/kubernetes/pull/91537 and https://github.com/kubernetes/kubernetes/pull/91525).

I figure that it makes the most sense to just drop this single conflicting line from the cherry-pick instead of 
Also cherry-picking the two (fairly large) standalone kube-aggregator OpenAPI generation PRs or
Abandoing the cherry-pick for 1.18 entirely.

Let me know if you’d prefer one of those options (or another one entirely)

kevindelgado commented:
/assign @apelisse 
/cc @liggitt

liggitt commented:
/ok-to-test

caesarxuchao commented:
/triage accepted

liggitt commented:
CI failures look related

kevindelgado commented:
I’ve fixed the CI issues that were causing the [1.20 backport](https://github.com/kubernetes/kubernetes/pull/100714) to fail, but the 1.18 and 1.19 backports are failing for a different reason.

The failing integration test `TestApplyStatus//v1,_Resource=services` gives the error:
```
=== RUN   TestApplyStatus//v1,_Resource=services
E0402 22:45:32.121024 2050397 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\""}
    status_test.go:162: failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field "protocol"
```

The error comes from this commit in the [cherry-pick here](https://github.com/kubernetes/kubernetes/pull/100714/commits/95714c2fe69d511dc41cad744e8aa0ab2af66e86#diff-b205a97d05b02829988fc305c6b78d5f272092e17f77fcbc78ca88c159648f99R169) where we modify the status test to use `rsc.Patch()` where it had previously used `rsc.Create()`. 

It wasn’t immediately clear why that change broke the test, but that error message led to [this issue](https://github.com/kubernetes-sigs/structured-merge-diff/issues/130) that mentions a [recent fix in 1.20](https://github.com/kubernetes-sigs/structured-merge-diff/issues/130#issuecomment-731697130). 

While the fix is not explicitly mentioned in the PR, I think @apelisse is talking about [this PR](https://github.com/kubernetes/kubernetes/pull/96480) that adds default fields to OpenAPI.

If that is the case, then one way to fix this would be to backport all the commits in that PR. It’s pretty hefty though and an initial attempt led me to a bunch of conflicts. Before I go down that rabbit hole and potentially hit other issues I wanted to check in with you two and see if we actually do want to backport the defaults PR to 1.18 and 1.19.

Alternatively, we could just get rid of the small commit that that changes the test to use patch instead of create and everything works fine. That’s what I’ve done here, just to check that it all passes CI, but let me know what route we want to take @apelisse @liggitt 

Whatever route we take, I’ll do the same to 1.19 unless you say otherwise (e.g I can see us wanting to go the easy way for 1.18 since the cherry pick deadline is in a week, but doing the full defaults backport for 1.19)

apelisse commented:
> If that is the case, then one way to fix this would be to backport all the commits in that PR. It’s pretty hefty though and an initial attempt led me to a bunch of conflicts. Before I go down that rabbit hole and potentially hit other issues I wanted to check in with you two and see if we actually do want to backport the defaults PR to 1.18 and 1.19.

Yeah, you can't do that. That feature can't be backported in older versions.

I also think that 693d0fb717be4840efcb404eab038070ce7f0881 should be omitted from this pull-request since it uses the feature implemented in #96480.

> The failing integration test `TestApplyStatus//v1,_Resource=services` gives the error:
> ```
> === RUN   TestApplyStatus//v1,_Resource=services
> E0402 22:45:32.121024 2050397 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\""}
>    status_test.go:162: failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field "protocol"
>```

You could also change this line:  https://github.com/kubernetes/kubernetes/blob/ce374017dc493aee5f64ae6239e2b3dc0267b99a/test/integration/etcd/data.go#L47
```diff
-Stub:             `{"metadata": {"name": "service1"}, "spec": {"externalName": "service1name", "ports": [{"port": 10000, "targetPort": 11000}], "selector": {"test": "data"}}}`,
+Stub:             `{"metadata": {"name": "service1"}, "spec": {"externalName": "service1name", "ports": [{"port": 10000, "targetPort": 11000, "protocol": "TCP"}], "selector": {"test": "data"}}}`,
```

And that would fix the problem you have applying this object.

apelisse commented:
Not changing the way we apply the object in the test would probably be fine too, that was mostly an improvement to the test that doesn't necessarily have to be carried into 1.18.

kevindelgado commented:
Sounds good, I've removed https://github.com/kubernetes/kubernetes/commit/693d0fb717be4840efcb404eab038070ce7f0881 from this and modified the etcd testdata to set the protocol instead of getting rid of the test improvement. I will do the same for the 1.19 backport.

Once this passes CI, PTAL @apelisse @liggitt

apelisse commented:
LGTM, thanks!

kevindelgado commented:
Friendly ping @liggitt (this needs to make it by the cherry-pick deadline on friday, since it's the last release of 1.18)

liggitt commented:
it's a little hard to tell what is being cherry-picked and what is new in this PR now. Can you update the title/description with the master commits that are now cherry-picked, and highlight the bits that are distinct to this PR (maybe just the protocol field in the etcd stub data?)

there also look to be formatting changes in several files that are likely related to running update-gofmt with a newer version of go. revert those if possible.

kevindelgado commented:
description updated, formatting updates removed, ptal @liggitt

kevindelgado committed:
update bazel

liggitt commented:
/retest
/lgtm
/approve

kevindelgado commented:
ready for cherry-pick review, ping @kubernetes/release-engineering

kevindelgado commented:
@liggitt can you tag this with /approved again? I think it didn't apply the label because it was still going through CI when you first applied it.

feiskyer commented:
@liggitt  could you approve it again? seems the original /approve comment didn't work

liggitt commented:
/approve
